pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,901 @@
+	}
+
+	result := false
+	for _, rule := range ua.config.Rules {
+		if rule.canTarget(AttachToExecutable) {
+			result = true
+			break
+		}
+	}
+	ua.handlesExecutablesCached = &result
+	return result
+}
+
+// Start starts the attacher, attaching to the processes and libraries as needed
+func (ua *UprobeAttacher) Start() error {
+	var cleanupExec, cleanupExit func()
+	procMonitor := monitor.GetProcessMonitor()
+	err := procMonitor.Initialize(ua.config.ProcessMonitorEventStream)
+	if err != nil {
+		return fmt.Errorf("error initializing process monitor: %w", err)
+	}
+
+	if ua.handlesExecutables() {
+		cleanupExec = procMonitor.SubscribeExec(ua.handleProcessStart)
+	}
+	// We always want to track process deletions, to avoid memory leaks
+	cleanupExit = procMonitor.SubscribeExit(ua.handleProcessExit)
+
+	if ua.config.PerformInitialScan {
+		// Initial scan only looks at existing processes, and as it's the first scan
+		// we don't have to track deletions
+		err := ua.Sync(true, false)
+		if err != nil {
+			return fmt.Errorf("error during initial scan: %w", err)
+		}
+	}
+
+	if ua.handlesLibraries() {
+		if !sharedlibraries.IsSupported(ua.config.EbpfConfig) {
+			return errors.New("shared libraries tracing not supported for this platform")
+		}
+
+		ua.soWatcher = sharedlibraries.NewEBPFProgram(ua.config.EbpfConfig)
+
+		err := ua.soWatcher.Init()
+		if err != nil {
+			return fmt.Errorf("error initializing shared library program: %w", err)
+		}
+		err = ua.soWatcher.Start()
+		if err != nil {
+			return fmt.Errorf("error starting shared library program: %w", err)
+		}
+	}
+
+	ua.wg.Add(1)
+	go func() {
+		processSync := time.NewTicker(ua.config.ScanProcessesInterval)
+
+		defer func() {
+			processSync.Stop()
+			if cleanupExec != nil {
+				cleanupExec()
+			}
+			cleanupExit()
+			procMonitor.Stop()
+			ua.fileRegistry.Clear()
+			if ua.soWatcher != nil {
+				ua.soWatcher.Stop()
+			}
+			ua.wg.Done()
+			log.Infof("uprobe attacher %s stopped", ua.name)
+		}()
+
+		var sharedLibDataChan <-chan *ebpf.DataEvent
+		var sharedLibLostChan <-chan uint64
+
+		if ua.soWatcher != nil {
+			sharedLibDataChan = ua.soWatcher.GetPerfHandler().DataChannel()
+			sharedLibLostChan = ua.soWatcher.GetPerfHandler().LostChannel()
+		}
+
+		for {
+			select {
+			case <-ua.done:
+				return
+			case <-processSync.C:
+				// We always track process deletions in the scan, to avoid memory leaks.
+				_ = ua.Sync(ua.config.EnablePeriodicScanNewProcesses, true)
+			case event, ok := <-sharedLibDataChan:
+				if !ok {
+					return
+				}
+				_ = ua.handleLibraryOpen(event)
+			case <-sharedLibLostChan:
+				// Nothing to do in this case
+				break
+			}
+		}
+	}()
+	log.Infof("uprobe attacher %s started", ua.name)
+
+	return nil
+}
+
+// Sync scans the proc filesystem for new processes and detaches from terminated ones
+func (ua *UprobeAttacher) Sync(trackCreations bool, trackDeletions bool) error {
+	if !trackDeletions && !trackCreations {
```

Comment:
can we have both flags set as `true` together?

Reply:
Yes, if the user sets `EnablePeriodicScanNewProcesses` to true that'd be the case. The istio monitor does this, I assume that's to be safe against lost events from the process monitor.